resource-managers/kubernetes/docker/src/main/dockerfiles/spark/Dockerfile

```diff
@@ -51,6 +51,8 @@ ENV SPARK_HOME /opt/spark
 
 WORKDIR /opt/spark/work-dir
 RUN chmod g+w /opt/spark/work-dir
+#Disable negative dns reslolution https://docs.oracle.com/javase/8/docs/technotes/guides/net/properties.html
+RUN sed -i -e 's/networkaddress.cache.negative.ttl=10/networkaddress.cache.negative.ttl=0/g'  /usr/lib/jvm/java-1.8-openjdk/jre/lib/security/java.security
```

Comment:
It would be more flexible if that could be set along with `[networkaddress.cache.ttl]`(https://docs.aws.amazon.com/sdk-for-java/v1/developer-guide/java-dg-jvm-ttl.html) via an env variable from within the entrypoint script.

Reply:
As I read in the docs, no those values can't be overriden via properties. It can be just replaced the whole file. 
https://stackoverflow.com/questions/4521119/java-argument-to-specify-java-security-file-for-jvm
You can use this code to test.

```
import java.security.Security;

public class HelloWorld {

    public static void main(String[] args) {
       String prop = Security.getProperty("networkaddress.cache.negative.ttl");
       System.out.println("java.security.networkaddress.cache.negative.ttl: " + prop);
    }
}
```

Reply:
I still find it better to be able to load your own file: `java.security.properties`. I dont agree with this being hardcode.

Reply:
I don't think we disable negative resolution caching anywhere. It has a potential negative impact, though, don't know if on the whole it's right. I'd leave it out of this otherwise minor change.

But, isn't the property just `networkaddress.cache.negative.ttl`? 
https://docs.oracle.com/javase/8/docs/api/java/net/InetAddress.html
At least that should be settable as a simple system property. Not sure if it's indeed different when trying to go this route.

Reply:
Hi @srowen @skonto 
The problem with Java and negative resolution is the moment you have ANY timeout in DNS resolution, you'll never be able to resolve that name. Which in some Kubernetes scenarios happens (pods network taking time to get fully operational). Basically no matter how many retries it will never work if you have a timeout.

Reply:
@skonto  About the options you said, I saw the option of sending the whole file,  I thought it was too complicated for people, and I think this issue is mostly present in K8s.

@srowen && @skonto  how do we continue with this?

Reply:
The default seems to be networkaddress.cache.negative.ttl = 10 or 10 seconds. Never caching might be better, but we don't do this anywhere else and seems like something that can be left to the user. Yeah OK it isn't a different property you're talking about and does seem only settable in the security props file, which isn't easy.
Never caching or caching for 10 seconds probably won't affect whether 1 or 3 tries fails or succeeds here, if that's the issue.

Reply:
Yes @srowen ,  exactly, it's 10 seconds. The issue comes with the actual look up which is a native invocation in jdk8, which has a timeout completely different to those 10 seconds, managed by the os.
So actually the code is not retrying at all because the results are cached. Unless the amount of retries would be absurdly high or we start sleeping in the code. In my opinion a retry limit of 10 seconds is way excessive. If you think about it the OS already provides this functionality.

Reply:
So, if we disable negative caching in the JVM, you mean the OS will still cache the negative resolution? does that leave the same problem, or do I misunderstand? are you just saying this is an attempt to get rid of the negative caching and may not succeed depending on OS?
Well, I'm not super against it if the K8S folks are OK with it, as this part is specific to K8S, but then, diverges from how it's run elsewhere, potentially.

Reply:
Hi @srowen 
Let me answer properly. AFAIK that DNS negative lookup is NOT cached, not at least in linux, positive caching yes, retries and timeouts, and it also depends on distribution, not all of them have it enabled (and there are different ways to achieve so).

What I meant is due to the fact that is cached, then retries won't work because they'll happen in a tight loop (the for loop we were adding), unless we start adding sleeps in the scala code, which may or may not be enough depending on the resolution timeouts at OS level. The end result of the caching of the negative resolution at java code is that final users end up with a very strange behaviour where the timeouts of the OS level are only respected once. 

What will happen if we don't disable caching, the first call to resolve, will wait the OS timeout (I think 5 seconds is the default), then the subsequent calls will just don't even try because the java layer won't try to invoke it. [here](https://github.com/bpupadhyaya/openjdk-8/blob/45af329463a45955ea2759b89cb0ebfe40570c3f/jdk/src/share/classes/java/net/InetAddress.java#L1251) and [here](https://github.com/bpupadhyaya/openjdk-8/blob/45af329463a45955ea2759b89cb0ebfe40570c3f/jdk/src/share/classes/java/net/InetAddress.java#L885)

So what this means is, if we don't disable negative DNS caches, then retries won't be able to solve the issue it in a reliable manner (a complicated combination of retries, sleeps in scala, and OS timers would be able to solve it).

Reply:
@jlpedrosa
>  I saw the option of sending the whole file, I thought it was too complicated for people, and I think this issue is mostly present in K8s.
I am not saying to send the whole the file, that is one option, user could just create it on the fly in his entrypoint script based on env vars. In my custom image for example I have a Prometheus config file that works for all my deployments, users in a slow env may even choose to have these security properties predefined in the file in the image. I dont see why you need to hardcoded them. 
The dockerfile you are modifying is about k8s only so we care about k8s here only.

Reply:
Hi @skonto 
For me any option is ok, the entrypoint for the container, I'll adapt to whatever you guys prefer. I am aware that the Dockerfile is only used in k8s, the train of thoughts is that this scenario is only common at k8s when pods networking may take some time (CNI, Sidecar pods..). Just let me know what is the path you preffer and I'll implement in such way.
In the docker image that spark uses (the base [spark image](https://github.com/apache/spark/blob/1217996f1574f758d8cccc1c4e3846452d24b35b/resource-managers/kubernetes/docker/src/main/dockerfiles/spark/Dockerfile#L18) is openjdk8 alpine [openjdk8](https://github.com/docker-library/openjdk/blob/178c542fbb93a8f8a42e331b73a1214c9d8ba81d/8/jdk/alpine/Dockerfile#L7))  dns caching is disabled, if you'd start an openjdk8 image and your try:
`watch -n 2 ping -c 1  www.apache.org`
while tcpdumping, you'll see that every 2 seconds a request is issued.
I tried this also in an ubtuntu docker image, and I get the same behaviour.

@skonto @srowen: Just let me know the way you'd prefer it and I'll patch it that way.

Reply:
Hi @jlpedrosa I voted for making it configurable and extendable ;) We should also make a note of this eg. docs? Let's see what @srowen has to say and has the final call.

Reply:
I just wouldn't make this change here.

Reply:
Hi @srowen,
Ok, then you'd you be ok to do the approach proposed by @skonto and have it as config option in the entrypoint script for the docker image?

Reply:
I'm neutral on that. It seems like a non-trivial change to the whole JVM to work around an env-specific DNS issue.

Reply:
Hi @srowen @skonto 

I've added the option as you considered it was acceptable. Please have a look and let me know what you think. This at least would enable to set it up via spark operator, but I can open a different PR if you want that I modify the code to manage that env variable in the scala side as a managed option for 2.4. In 3.0 it can be handled via pod templates, have not tested it, but I think all scala is doing append variables, so defining that variable in a template could enable it